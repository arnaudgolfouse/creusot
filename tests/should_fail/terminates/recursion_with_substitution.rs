// Ensures that variants are correctly checked when recursing while substituing some types.

extern crate creusot_contracts;
use creusot_contracts::{logic::WellFounded, *};

pub trait MyTrait {
    #[logic]
    fn as_int(self) -> u32;

    #[requires(self.as_int()@ - x@ >= 0)]
    #[ensures(result@ == self.as_int()@ - x@)]
    #[check(terminates)]
    fn sub(self, x: u32) -> u32;
}
impl MyTrait for u32 {
    #[logic(open)]
    fn as_int(self) -> u32 {
        self
    }

    #[check(terminates)]
    #[requires(self@ - x@ >= 0)]
    #[ensures(result@ == self@ - x@)]
    fn sub(self, x: u32) -> u32 {
        self - x
    }
}

#[allow(unconditional_recursion)]
#[variant(arg_0)]
#[check(terminates)]
#[ensures(result == 0u32)]
pub fn foo<T: WellFounded + MyTrait>(arg_0: T) -> u32 {
    foo(0)
}
