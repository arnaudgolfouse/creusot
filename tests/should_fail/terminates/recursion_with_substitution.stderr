error: Polymorphic recursion is not supported: recursive calls should have the same type parameters.
  --> recursion_with_substitution.rs:35:5
   |
35 |     foo(0)
   |     ^^^^^^

error: aborting due to 1 previous error

